Add destroy server group ops and other minor bug fixes

milismsft opened the pull request:
Add destroy server group ops. This payload was reviewed previously; the only addition is the test and removing some commented out lines in the validation class.
Add type for the deployment. Updating the python tests to reflect these changes is a separate task.
Cleanup resources that are created as part of the create server group in case of a failure during the creation.
Add retries support when deleting a resource in Azure (work around for random timeouts).
Fix an assertion in clouddriver console when the application has no clusters.

@JargoonPard @scotmoor PTAL; I'm looking to submit A PR with these changes later tonight.

milismsft committed:
Add destroy server group ops and other minor bug fixes

ajordens committed:
rx-ified the ReservationReportCachingAgent

- Each account will be loaded in parallel

scotmoor commented on clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureComputeClient.groovy:
```diff
@@ -154,6 +155,28 @@ public class AzureComputeClient extends AzureBaseClient {
+        operationRetry ++
+        result = this.client.getVirtualMachineScaleSetsOperations()?.delete(resourceGroupName, serverGroupName)
+        operationRetry = AZURE_ATOMICOPERATION_RETRY
+      } catch (Exception e) {
```
What if the exception is a cloud exception and the status code is a 404? Would you want to keep trying to delete it if it's already gone?

milismsft replied:
I checked the Azure JSDK implementation and for post or delete operations they don't handle 404:
com.microsoft.azure/azure-client-runtime/1.0.0-beta1/79374d64473fc93e2a2478e7b374dd5d289c98b1/azure-client-runtime-1.0.0-beta1-sources.jar!/com/microsoft/azure/AzureClient.java:270
        if (statusCode != 200 && statusCode != 202 && statusCode != 204) {
            CloudException exception = new CloudException(statusCode + " is not a valid polling status code");
            exception.setResponse(response);

scotmoor commented on clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureNetworkClient.groovy:
```diff
@@ -184,13 +184,61 @@ class AzureNetworkClient extends AzureBaseClient {
+        operationRetry ++
+        result = client.getLoadBalancersOperations().delete(resourceGroupName, loadBalancerName)
+        operationRetry = AZURE_ATOMICOPERATION_RETRY
+      } catch (Exception e) {
```
Same comment about handling 404s

ajordens committed:
Merge pull request #487 from ajordens/optimized-reservation-report-caching-agent

rx-ified the ReservationReportCachingAgent

JargoonPard commented:
LGTM

Commit:
Fixes issue with order of GCE httpHealthCheck and instance health check calls.

Commit:
Merge pull request #488 from ttomsu/google-redis-27

Fixes issue with order of GCE httpHealthCheck and instance health che…

Commit:
provider/kubernetes: Run Job operation

Commit:
provider/docker: Defer loading _catalog until after initialization

ajordens committed:
Introduced a ReservedInstances caching agent (runs every 10 minutes)

- Updated ReservationReport caching agent to use the cached ReservedInstances data rather than calling AWS directly
- Adjusted ReservationReport caching agent to only run every minute

ajordens committed:
No need to @Autowire

ajordens committed:
Merge pull request #492 from ajordens/reserved-instances-caching-agent

Introduced a ReservedInstances caching agent (runs every 10 minutes)

lwander committed:
Merge pull request #491 from lwander/docker-defer-catalog

provider/docker: Defer loading _catalog until after initialization

tomaslin committed:
titus: handle case where ports can be null

tomaslin committed:
Merge pull request #494 from tomaslin/titus-port-can-be-empty

titus: handle case where ports can be null

cfieber committed:
filter instance relationships for significant performance improvement on large server groups

ajordens committed:
Merge pull request #495 from cfieber/fetch_sg_perf

filter instance relationships for significant performance improvement on large server groups

ajordens committed:
Re-generate block device mappings if instance type changes

- Only applies if the source ASG was using the default mappings for the source instance type
- Include duration in cache agent completion logging (for convenience more than anything)

ajordens committed:
AgentSchedulers are now responsible for tracking elapsed execution time

- No more ThreadLocal in MetricInstrumentation

ajordens committed:
Merge pull request #496 from ajordens/fix-block-device-mappings-when-cloning-asg

Fix block device mappings when cloning asg

Commit:
Fix the setting of the serverGroup.disabled property to take into account whether there are load balancers defined in the server group metadata.

Commit:
Merge pull request #497 from duftler/google-fix-disabled

Fix the setting of the serverGroup.disabled property to take into acc…

tomaslin committed:
moves the health providing caching agent from aws to core

anotherchrisberry committed:
Merge pull request #489 from Microsoft/azure-destroysg-pr

Add destroy server group ops and other minor bug fixes

tomaslin committed:
expose titus entrypoints

tomaslin committed:
Merge pull request #500 from tomaslin/titus-command

titus: expose endpoints in api

Commit:
instance caching bug fix

# Conflicts:
#	clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureBaseClient.groovy

tomaslin committed:
Merge pull request #493 from Microsoft/cachebugfix

instance caching bug fix

ajordens committed:
ReservedInstances caching agent only caches 'active' RIs

- Additional logging in ReservationReport caching agent

ajordens committed:
Merge pull request #501 from ajordens/only-cache-active-reserved-instances

ReservedInstances caching agent only caches 'active' RIs

ajordens committed:
Expose maxConnections and maxConnectionsPerRoute as configurable options used by the AmazonClientProvider

- Defaults: { maxConnections: 200, maxConnectionsPerRoute: 20 }

ajordens committed:
Merge pull request #502 from ajordens/expose-maxConnections-and-maxConnectionsPerRoute

Expose maxConnections and maxConnectionsPerRoute as configurable opti…

cfieber committed:
Adds NodeStatusProvider for ClusteredAgentScheduler.

ClusteredAgentScheduler will only run agents if the node is enabled.

Adds EurekaStatusNodeStatusProvider to determine node status from eureka status.

cfieber committed:
Merge pull request #503 from cfieber/disable_caching_agents_when_oos

Adds NodeStatusProvider for ClusteredAgentScheduler.

Commit:
provider/kubernetes: Fix image pull secret address

Commit:
Merge pull request #504 from lwander/kubernetes-correct-creds

provider/kubernetes: Fix image pull secret address

lwander committed:
Merge pull request #490 from lwander/kubernetes-jobs

provider/kubernetes: Run Job operation

Commit:
Fix deserialization error when moving from on demand namespace.

Commit:
Merge pull request #505 from ttomsu/google-redis-28

Fix deserialization error when moving from on demand namespace.

Commit:
Consider serverGroup.disabled when returned server group disabled/enabled status as part of load balancer view.

Commit:
Merge pull request #506 from duftler/google-fix-lb-disabled

Consider serverGroup.disabled when returned server group disabled/ena…

tomaslin committed:
Merge pull request #499 from tomaslin/move-caching-agent-to-core

moves the HealthProvidingCachingAgent from aws to core
